src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs

```diff
@@ -231,11 +238,46 @@ protected BoundCatchBlock GenerateExceptionHandling(LocalSymbol exceptionLocal)
                 exceptionFilterOpt: null,
                 body: F.Block(
                     assignFinishedState, // _state = finishedState;
+                    GenerateHoistedLocalsCleanup(hoistedLocals),
                     callSetException, // builder.SetException(ex);  OR  _promiseOfValueOrEnd.SetException(ex);
                     GenerateReturn(false)), // return;
                 isSynthesizedAsyncCatchAll: true);
         }
 
+        protected BoundStatement GenerateHoistedLocalsCleanup(ImmutableArray<StateMachineFieldSymbol> hoistedLocals)
+        {
+            var builder = ArrayBuilder<BoundStatement>.GetInstance();
+
+            // Cleanup all hoisted local variables
+            // so that they can be collected by GC if needed
+            foreach (var hoistedLocal in hoistedLocals)
+            {
+                if (!CanReferenceHeap(hoistedLocal.Type))
+                {
+                    continue;
+                }
+
+                builder.Add(F.Assignment(F.Field(F.This(), hoistedLocal), F.NullOrDefault(hoistedLocal.Type)));
+            }
+
+            return F.Block(builder.ToImmutableAndFree());
+        }
+
+        private bool CanReferenceHeap(TypeSymbol type)
+        {
+            if (type.SpecialType.IsValueType())
```

Comment:
Value types can reference the heap by containing fields of reference type.  The simplest way to see if a value type can reference the heap is to check if the type is unmanaged, in which case it cannot reference the heap.

Also, a tuple type is a value type, so the if statement below will never succeed.

Reply:
Thanks for the remark, fixed.

But Tuple is not a SpecialType, isn't it? That's why I previously handle it separately.

Reply:
Can using IsManaged lead to a problem in case a type is unmanaged in compilation time but loaded from another assembly and become managed in runtime?

Reply:
> Also I suppose that using IsManaged may lead to a problem in case a type is unmanaged in compilation time but loaded from another assembly and becomes managed in runtime.

I don't think you need to worry about this. It's essentially the same problem as if we took a pointer to such an unmanaged type. Changing a struct's fields between compilation and runtime is basically a breaking change, I don't know what the runtime does but I would not expect it to work.

Reply:
Why changing a struct's fields without recompilation may break anything in runtime? All the IL code reference fields with tokens so any change is handled by the runtime.

Still, considering the code I've added slows down every async method invocation I totally agree it's better to have better performance in general case over better performance in some tricky edge case.

Reply:
> Also I suppose that using `IsManaged` may lead to a problem in case a type is unmanaged in compilation time but loaded from another assembly and becomes managed in runtime.

When you say "a problem", you mean the same problem we have with every type today (not cleared when state machine exits)?  That still seems to be a much better place than where we are today.

Reply:
> When you say "a problem", you mean the same problem we have with every type today (not cleared when state machine exits)?
Yep.

> That still seems to be a much better place than where we are today.
I couldn't agree more.